lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1993,7 +2036,188 @@ LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
     }
   }
   rewriter.eraseOp(op);
-  return success();
+  return true;
+}
+
+// If memory has no write ports, eliminate it.
+bool foldReadOrWriteOnlyMemory(PatternRewriter &rewriter, MemOp op) {
+  bool isRead = false, isWritten = false;
+  for (unsigned i = 0; i < op.getNumResults(); ++i) {
+    switch (op.getPortKind(i)) {
+    case MemOp::PortKind::Read:
+      isRead = true;
+      if (isWritten)
+        return false;
+      continue;
+    case MemOp::PortKind::Write:
+      isWritten = true;
+      if (isRead)
+        return false;
+      continue;
+    case MemOp::PortKind::Debug:
+    case MemOp::PortKind::ReadWrite:
+      return false;
+    }
+    llvm_unreachable("unknown port kind");
+  }
+  assert((!isWritten || !isRead) && "memory is in use");
+
+  for (auto port : op.getResults()) {
+    auto dummyWire = rewriter.create<WireOp>(port.getLoc(), port.getType());
```

Comment:
Does this replace memories with no write ports with uninitialized wires ?

Reply:
Yes. Any better ideas on what to do in such a case? I noticed that the wires are not optimized away properly after `LowerToHW`.

Reply:
We can handle un-initialized wires specially in `LowerToHW`. I think `LowerToHW` was not expecting such wires, so its not handled.